tests/test_arithmetics.rs

```diff
@@ -423,3 +423,193 @@ fn rich_comparisons_python_3_type_error() {
+        }
+
+        fn __richcmp__(&self, other: PyRef<'p, Self>, _op: CompareOp) -> PyObject {
+            other.py().None()
+        }
+    }
+
+    #[pyproto]
+    impl<'p> PyNumberProtocol<'p> for RichComparisonToSelf {
+        fn __add__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __sub__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __mul__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __matmul__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __truediv__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __floordiv__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __mod__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __lshift__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __rshift__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+        fn __divmod__(lhs: &'p PyAny, _other: PyRef<'p, Self>) -> &'p PyAny {
+            lhs
+        }
+
+        fn __iadd__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __isub__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __imul__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __imatmul__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __itruediv__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __ifloordiv__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __imod__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __ipow__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __ilshift__(&'p mut self, _other: PyRef<'p, Self>) {}
+        fn __irshift__(&'p mut self, _other: PyRef<'p, Self>) {}
+    }
+
+    fn test_dunder(operator: &str, raises_typeerror: bool) {
+        let gil = Python::acquire_gil();
+        let py = gil.python();
+        let c2 = PyCell::new(py, RichComparisonToSelf {}).unwrap();
+        py_run!(
+            py,
+            c2,
+            &format!(
+                "\
+class Other:
+    def __eq__(self, other):
+        return True
+    __ne__ = __lt__ = __le__ = __gt__ = __ge__ = __eq__
+
+    def __radd__(self, other):
+        return self
+    __rand__ = __ror__ = __rxor__ = __radd__
+    __rsub__ = __rmul__ = __rtruediv__ = __rfloordiv__ = __rpow__ = __radd__
+    __rmatmul__ = __rlshift__ = __rrshift__ = __rmod__ = __rdivmod__ = __radd__
+
+c2 {} Other()",
+                operator
+            )
+        );
+
+        if raises_typeerror {
+            py_expect_exception!(
+                py,
+                c2,
+                &format!("class Other: pass\nc2 {} Other()", operator),
+                PyTypeError
+            );
+        }
+    }
+
+    macro_rules! not_implemented_test {
```

Comment:
I want to avoid using this macro to generate test *functions*. It makes it more difficult to maintain our tests.
What is important is to place each test with clear purpose.

Reply:
I disagree here; I tend to think that if it's possible to write a simple macro which generates tests to cover the entire set expected of expected behaviour, then it's easier to maintain that rather than many individual tests to cover all the interface.

But I accept this is a style choice, so willing to cede to your preference @kngwyu

Reply:
Hmm :thinking: 
It can be true sometimes but in this case, what we do is simply calling `test_dunder`.
So I think simply repeating `test_dunder` is not so bad.
```rust
#[test] 
fn test_operators_return_not_implemented() {
    test_impl("<");
    test_impl(">");
    ....
}
```

Reply:
See https://github.com/PyO3/pyo3/pull/1072#discussion_r464069161.